src/program/latex/diagram/mod.rs

```diff
@@ -0,0 +1,174 @@
+                    }
+                }
+            }
+        }
+
+        // get display of gate name from composite gate or original gate
+        let canonical_gate = CompositeGate::from_str(&gate.name)
+            .map(|g| g.to_string())
+            .unwrap_or(gate.name.clone());
+
+        // get the names of the qubits in the circuit before circuit is borrowed as mutable
+        let circuit_qubits: Vec<u64> = self.circuit.keys().cloned().collect();
+
+        // set gate for each qubit in the instruction
+        for qubit in &gate.qubits {
+            if let Qubit::Fixed(instruction_qubit) = qubit {
+                if let Some(wire) = self.circuit.get_mut(instruction_qubit) {
+                    // set the control and target qubits
+                    if gate.qubits.len() > 1 || canonical_gate == "PHASE" {
+                        // set the target qubit if the qubit is equal to the last qubit in gate
+                        if qubit == gate.qubits.last().unwrap() {
+                            wire.set_targ();
+                        // otherwise, set the control qubit
+                        } else {
+                            wire.set_ctrl(qubit, gate.qubits.last().unwrap(), &circuit_qubits);
+                        }
+                    } else if wire.parameters.get(&wire.column).is_some() {
+                        // parameterized single qubit gates are unsupported
+                        return Err(LatexGenError::UnsupportedGate {
+                            gate: gate.name.clone(),
+                        });
+                    }
+
+                    // set modifiers at this column for all qubits
+                    wire.gates.insert(wire.column, canonical_gate.clone());
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl fmt::Display for Diagram {
+    /// Returns a result containing the body of the Document.
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // write a newline between the body and the Document header
+        writeln!(f)?;
+
+        // write the LaTeX string for each wire in the circuit
+        for (qubit, wire) in &self.circuit {
+            // are labels on in settings?
+            if self.settings.label_qubit_lines {
+                // write the label to the left side of wire
+                write!(f, "{}", RenderCommand::Lstick(*qubit))?;
+            } else {
+                // write an empty column buffer as the first column
+                write!(f, "{}", RenderCommand::Qw)?;
+            }
+
+            // write the LaTeX string for the wire
+            write!(f, "{wire}")?;
+
+            // chain an empty column to the end of the line
+            write!(f, " & ")?;
+            write!(f, "{}", &RenderCommand::Qw)?;
+
+            // if this is the last iteration, omit a new row from the end of the line
+            if *qubit != *self.circuit.keys().last().unwrap() {
```

Comment:
is it worth using `for (i, (qubit, wire)) in &self.circuit.enumerate() {` or similar to make the check here simpler? Or cache `let last = self.circuit.keys().last().unwrap()` before the loop? Second one might need extra handling if `self.circuit.len() == 0`.

Not a blocker, just curious other's thoughts

Reply:
Similar to the last previous review I was able to pull it outside of the for-loop in general. I was able to handle the len==0 case with, `let last = self.circuit.keys().last().unwrap_or(&0);` Now it only needs to run once.